drivers: entropy: fix native_posix driver for more than 3 byte requests

aescolar opened the pull request:
Fix the native_posix fake entropy driver for more than 3 byte requests, and specially for native_sim//64 builds.

The host random() provides a number between 0 and 2**31-1 (INT_MAX), so bit 32 was always 0.
So when filling a buffer with more than 3 bytes we would be filling each 4th byte with a byte which always had its MSbit as 0.

For LP64 native_sim//64 builds, this was even worse, as the driver had another bug where we assumed random() returned the whole long filled, and therefore all 4 upper bytes would be 0.

Fixes #83447

aescolar committed:
drivers: entropy: fix native_posix driver for more than 3 byte requests

Fix the native_posix fake entropy driver for more than 3 byte requests,
and specially for native_sim//64 builds.

The host random() provides a number between 0 and 2**31-1 (INT_MAX),
so bit 32 was always 0.
So when filling a buffer with more than 3 bytes we would be filling
each 4th byte with a byte which always had its MSbit as 0.

For LP64 native_sim//64 builds, this was even worse, as the driver had
another bug where we assumed random() returned the whole long filled,
and therefore all 4 upper bytes would be 0.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

tomi-font commented on drivers/entropy/fake_entropy_native_posix.c:
```diff
@@ -35,18 +36,25 @@ static int entropy_native_posix_get_entropy(const struct device *dev,
 		 */
-		long value = nsi_host_random();
+		value = value | ((uint64_t)nsi_host_random() << bits_ready);
+		/* The host random() provides a number between 0 and 2**31-1. Bit 32 is always 0 */
+		bits_ready += 31;
 
-		size_t to_copy = MIN(length, sizeof(long int));
+		size_t to_copy = MIN(length, bits_ready / 8);
 
 		memcpy(buffer, &value, to_copy);
 		buffer += to_copy;
 		length -= to_copy;
+		value >>= to_copy * 8;
+		bits_ready -= to_copy * 8;
```
I find this a bit complicated. Why not just copy 3 bytes at a time at every loop iteration and dump the remaining 7 bits returned by the host random? I don't think we're concerned about performance to that point here?

aescolar replied:
Fair enough :), I did overdo it indeed.

tomi-font reviewed (commented):
Literally fake entropy it is. :smile:

aescolar committed:
tests/subsys/lorawan/frag_decoder: Change random seed

These tests are quite sensitive to the exact random sequence.
After a change of the native_sim entropy generation 2 of them
started failing. Let's set a random seed which avoids the failure.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
